Adds NBT data to the blocks to handle unlocalized names for block textures.

AnvilPuncher opened the pull request:
Hi,
In this fork, the readFromNBT and writeToNBT functions in TEBase.class were modified to add and handle an extra string to the cbAttributes tag called "Texture".  The short, "id", is still used and it is able to handle the special textures from seeds, wheat, mycellium, etc.  We did a little stress testing to accommodate the case when the unlocalized name in "Texture" did not represent an existing block and also when the "id" number did not represent an existing block as well.  
The code probably isn't as elegant as what it could be but it seems to work. 
With these changes, if a user ports a saved world to another environment that has different mods, the textures of the blocks should not be directed by the id number of the block but rather the unlocalized name.  If in case the unlocalized name isn't an existing block in that environment, rather than throwing an exception, it simply clears out the data in "cbAttrList" tag.  

I hope this can help anyone.

Mineshopper commented:
I don't currently have my environment set up, but when I do I will look into how Forge is handling the ItemStacks to see if this solution makes sense.  Or, if you wanted to peek at how it resolves the block from an ItemStack, you might beat me to the punch.  I basically want to see if it resolves the block based on an ID (which I thought were discarded in 1.7), or whether it uses the unlocalized name.

As for your code changes, I noticed that the attribute list is discarded entirely under certain situations, but many of the attributes don't contribute to the appearance of the block.  The covers are what "texture" the block and all six protruding sides, and overlays are drawn on top of those.  Cover ID 6 is the one you need to focus on.

I can't give an ETA on getting my environment set up, but I will try to do it relatively soon.

AnvilPuncher commented:
Yeah, it looks like ItemStack uses the numerical ID.  That is what it is putting in the NBT data tag.  Of course, this means that if the ID is not of a vanilla ID, then it will be more or less unique for that world.  T

There is code that removes the attribute list entirely in the case that the unlocalized name found in "Texture" isn't a known block.  It would otherwise throw an exception.  For Cover ID 0 through 5, these also need to be handled in a similar fashion because they too hold a numerical ID for the block that it is getting its texture from.  I figured that it would be better for the block to revert to an untextured state rather than throw an exception in forge.  

This code is pretty much to enforce that the numerical ID is what the builder originally intended by adding the "Texture" data tag.  The numerical ID will be overwritten if it differs from what the level.dat information pairs it with the unlocalized name.  

Thanks

Mineshopper commented:
I see what you're talking about now.  I will integrate with next stable build.

Mineshopper commented:
Do you have an example where texture data is lost in the current development build?  I am adapting this to a more comprehensive solution but am not sure when block attributes are lost.  I have tried loading a couple mods, and removing one.. but attributes for the remaining mod seem to remain.

AnvilPuncher commented:
Texture data isn't lost in the current dev build.  I added the code to handle situations that can arise in what I had added.  That situation is when the block that the carpenter's texture points to does not exist.  It is only a real problem now since the data contains unlocalized data name that is getting checked against the block registry.  If it doesn't exist, then it throws an exception in forge.  I made it simply remove the blocks texture data entirely rather than throw exceptions in that case.

Mineshopper commented:
Sorry, I meant prior to your fix.  What situations before resulted in texture data being lost?

AnvilPuncher commented:
The texture data was never lost.  The data simply pointed to the ID number of the texture block inside of the carpenter's block.  The numerical ID of a block (outside of vanilla) is not set in stone.  If you load a mod that makes a block with an ID of, for example, 2345 and then remove the mode, you are still left with that block (in that world) with a texture reference to a block ID 2345.  This reference number may not exist or exist with another block at this point.  Also, there are mods that capable of constructing structures, other other things, with blocks with NBT data.  If someone makes a structure containing a carpenter's block that had a texture ID of 2345, because it gave a certain texture, and then sent that data to another person there exist a good possibility that the texture on that block wouldn't be the same.  
     Unfortunately, the situation in carpenter's blocks isn't a new one.  I believe there are other mods that allow textures to be added, such as malises mods, bibliocraft, storage drawers.  I know that bibliocraft has a unique way of circumventing the ID issue.  It just has a NBT tag that points directly to the name of the block's PNG graphic.  Which can work but it assumes all 6 faces are the same.  Malises (various mods) and storage drawers use the ID number system from ItemStack.  Because of this, their stuff runs into the same problem as yours had.  
     In my best answer for your question, I came upon this problem while creating structures to be used in the mod "Chocolate Quest."  I would make a structure using carpenter's blocks then create the "schematic" using Chocolate Quest's schematic builder.  It would build the schematic and record all the NBT data for the blocks therein.  I sent the schematic to my wife for some testing.  She was running an environment that had one extra mod than I did.  This mod added about 6 or so blocks and Forge just so happened to put them numerically lower than most other blocks.  This pushed the ID of most other blocks by 6 more numbers.  When she built the schematic and all of the textures inside the carpenter's blocks where different from what I did.  It was reasonable what happened but at the time we didn't know what carpenter's blocks used to retain this data.  This spurred the fork and pull request.  

Thank you for your attention to all of this and I'll help in anyway I can.
